src/Hl7.Fhir.Support.Poco/ElementModel/ScopedNodeExtensions.cs

```diff
@@ -0,0 +1,115 @@
+                }
+
+                // Return the identity - will remain relative if we did not find a fullUrl              
+            }
+
+            return identity;
+        }
+
+        public static T Resolve<T>(this T element, string reference, Func<string, T> externalResolver = null) where T : class, ITypedElement
+        {
+            // Then, resolve the url within the instance data first - this is only
+            // possibly if we have a ScopedNode at hand
+            if (element is ScopedNode scopedNode)
+            {
+                var identity = scopedNode.MakeAbsolute(new ResourceIdentity(reference));
+
+                if (identity.IsLocal || identity.IsAbsoluteRestUrl || identity.IsUrn)
+                {
+                    var result = locateResource(identity);
+                    if (result != null) return (T)(object)result;
+                }
+            }
+
+            // Nothing found internally, now try the external resolver
+            if (externalResolver != null)
+                return externalResolver(reference);
+            else
+                return null;
+
+            ScopedNode locateResource(ResourceIdentity identity)
+            {
+                var url = identity.ToString();
+
+                foreach (var parent in scopedNode.ParentResources())
+                {
+                    if (parent.InstanceType == "Bundle")
+                    {
+                        var result = parent.BundledResources().FirstOrDefault(br => br.FullUrl == url)?.Resource;
+                        if (result != null) return result;
+                    }
+                    else
+                    {
+                        if (parent.Id() == url) return parent;
+                        var result = parent.ContainedResources().FirstOrDefault(cr => cr.Id() == url);
+                        if (result != null) return result;
+                    }
+                }
+
+                return null;
+            }
+        }
+
+        /// <summary>
+        /// Where this item is a reference, resolve it to an actual resource, and return that
+        /// </summary>
+        /// <param name="element"></param>
+        /// <param name="externalResolver"></param>
+        /// <returns></returns>
+        public static T Resolve<T>(this T element, Func<string, T> externalResolver = null) where T : class, ITypedElement
+        {
+            // First, get the url to fetch from the focus
+            string url = null;
+
+            if (element.InstanceType == "string" && element.Value is string s) // hmm, we cannot use FhirAllTypes.String.GetLiteral()
+                url = s;
+            else if (element.InstanceType == "Reference")  // hmm, we cannot use FhirAllTypes.Reference.GetLiteral()
+                url = element.ParseResourceReference()?.Reference;
+
+            if (url == null) return default;   // nothing found to resolve
```

Comment:
Yes, your question is a good one.  The agnostic bits use magic string for the most common datatypes/resources.  This will certainly improve when we are using ModelSpace, but when is that?  Should we jump and simple introduce some constants?  Where ?  I guess in Hl7.Fhir.Support.Poco?